### PR TITLE
Enable all endpoint functions in the class hierarchy

### LINKED ISSUE #7865: dynamicReward_getDefaultRewardAtHeight endpoint is not exposed
### Expected behavior
`dynamicReward_getDefaultRewardAtHeight` endpoint should be available on lisk-core start.

### Actual behavior
`dynamicReward_getDefaultRewardAtHeight` endpoint is not exposed.

### Steps to reproduce
Start lisk-core using branch `release/4.0.0`. Will get following response from the above endpoint.
```
error: {
    message: 'Request 'dynamicReward_getDefaultRewardAtHeight' is not registered to bus.'
}
```

### Which version(s) does this affect? (Environment, OS, etc...)
lisk-sdk: 6.0.0-alpha.7
lisk-core: 4.0.0-alpha.5

### PR DESCRIPTION
**Pull request by bobanm**
### What was the problem?

This PR resolves #7865

This might not be the best solution. I'm pushing this as a draft PR to discuss the right approach.

### How was it solved?

`DynamicRewardEndpoint` module is unique as unlike the other endpoints, it extends `RewardEndpoint` instead of `BaseEndpoint`. In the future, there could be other endpoint classes with similar behavior.

When getting endpoint handlers, the function now loops through the prototype hierarchy until it reaches `Object` and collects all the endpoint functions along the way.

### How was it tested?

When running the example app, all endpoints are now correctly handled on my machine. But there are quite a few tests which fail now. I think they fail for 2 reasons:

1. The test data is no longer compatible and should be made more realistic
2. Some tests use `LegacyEndpoint` whose structure is not covered by the changes implemented

### CODE REVIEW

File: framework/src/endpoint.ts [hunk 1]
```diff
@@ -12,16 +12,28 @@
  * Removal or modification of this copyright notice is prohibited.
  */
 
+import { BaseEndpoint } from './modules';
+import { DynamicRewardEndpoint } from './modules/dynamic_rewards/endpoint';
```
Comment (shuse2) [inline framework/src/endpoint.ts:16]: Framework in general should not be aware of any module.
If we need specific handling, we need to consider different approach in the module side

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: framework/src/endpoint.ts [hunk 2]
```diff
@@ -21,7 +21,18 @@ export const isReservedEndpointFunction = (key: string): boolean =>
...
+	// For endpoints which extend other endpoints, go deeper in the prototype chain
+	do {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, no-param-reassign
+		endpoint = Object.getPrototypeOf(endpoint);
+		isBaseEndpoint =
+			(Object.getPrototypeOf(Object.getPrototypeOf(endpoint)) as Record<string, unknown>)
+				.constructor.name === Object.name;
+	} while (!isBaseEndpoint);
+
+	for (const key of Object.getOwnPropertyNames(endpoint)) {
```
Comment (shuse2) [inline framework/src/endpoint.ts:35]: i think in this way, it's not possible to get both base endpoint and some endpoint on the extended class?

File: framework/src/endpoint.ts [hunk 3]
```diff
@@ -21,12 +21,25 @@ export const isReservedEndpointFunction = (key: string): boolean =>
...
-	for (const key of Object.getOwnPropertyNames(Object.getPrototypeOf(endpoint))) {
-		const val = endpoint[key];
-		if (!isReservedEndpointFunction(key) && typeof val === 'function') {
-			endpointHandlers[key] = val.bind(endpoint) as EndpointHandler;
+	let isBaseEndpoint;
+
+	// For endpoints which extend other endpoints, go deeper in the prototype chain
+	do {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, no-param-reassign
+		endpoint = Object.getPrototypeOf(endpoint);
```
Comment (shuse2) [inline framework/src/endpoint.ts:29]: Prefer to avoid `no-param-reassign`

File: framework/src/endpoint.ts [hunk 4]
```diff
@@ -21,12 +21,25 @@ export const isReservedEndpointFunction = (key: string): boolean =>
...
+			const val = endpoint[key];
+			if (!isReservedEndpointFunction(key) && typeof val === 'function') {
+				endpointHandlers[key] = val.bind(endpoint) as EndpointHandler;
+			}
 		}
-	}
+
+		isBaseEndpoint =
+			(Object.getPrototypeOf(Object.getPrototypeOf(endpoint)) as Record<string, unknown>)
+				.constructor.name === Object.name;
```
Comment (shuse2) [inline framework/src/endpoint.ts:40]: why is this comparing with `Object.name`('Object')?

Comment (bobanm) [reply framework/src/endpoint.ts:40]: What would be a more elegant way to detect that the loop has reached the end where the prototype returned by `Object.getPrototypeOf()` is `Object`?

**Review by shuse2 — APPROVED** [review]
LGTM but `yarn.lock` is giving the change probably from development branch.

**Comment by bobanm** [comment]
> LGTM but `yarn.lock` is giving the change probably from development branch. 

@shuse2 what do you mean?

**Comment by shuse2** [comment]
@bobanm  CI is failing on the `format` step

**Review by ishantiw — APPROVED** [review]

**Comment by shuse2** [comment]
@bobanm it should be fixed now so disregard my comment =)

**Pull request merged**